eval/src/vespa/eval/instruction/generic_peek.cpp

```diff
@@ -0,0 +1,364 @@
+
+    ~SparsePlan();
+
+    template <typename Getter>
+    SparseState make_state(const Getter &get_child_value) const {
+        std::vector<vespalib::string> view_addr;
+        auto vpos = verbatim.begin();
+        auto cpos = children.begin();
+        for (auto source : sources) {
+            switch (source) {
+            case Source::VERBATIM:
+                view_addr.push_back(*vpos++);
+                break;
+            case Source::FROM_CHILD:
+                view_addr.push_back(vespalib::make_string("%" PRId64, get_child_value(*cpos++)));
+                break;
+            }
+        }
+        assert(vpos == verbatim.end());
+        assert(cpos == children.end());
+        assert(view_addr.size() == view_dims.size());
+        return SparseState(std::move(view_addr), out_mapped_dims);
+    }
+};
+SparsePlan::~SparsePlan() = default;
+
+struct PeekParam {
+    const ValueType res_type;
+    DensePlan dense_plan;
+    SparsePlan sparse_plan;
+    size_t num_children;
+    const ValueBuilderFactory &factory;
+
+    PeekParam(const ValueType &input_type,
+              const ValueType &res_type_in,
+              const GenericPeek::SpecMap &spec_in,
+              const ValueBuilderFactory &factory_in)
+        : res_type(res_type_in),
+          dense_plan(input_type, spec_in),
+          sparse_plan(input_type, spec_in),
+          num_children(count_children(spec_in)),
+          factory(factory_in)
+    {
+        assert(dense_plan.in_dense_size == input_type.dense_subspace_size());
+        assert(dense_plan.out_dense_size == res_type.dense_subspace_size());
+    }
+};
+
+template <typename ICT, typename OCT, typename Getter>
+Value::UP
+generic_mixed_peek(const ValueType &res_type,
+                   const Value &input_value,
+                   const SparsePlan &sparse_plan,
+                   const DensePlan &dense_plan,
+                   const ValueBuilderFactory &factory,
+                   const Getter &get_child_value)
+{
+    auto input_cells = input_value.cells().typify<ICT>();
+    size_t bad_guess = 1;
+    auto builder = factory.create_value_builder<OCT>(res_type,
+                                                     sparse_plan.out_mapped_dims,
+                                                     dense_plan.out_dense_size,
+                                                     bad_guess);
+    size_t filled_subspaces = 0;
+    size_t dense_offset = dense_plan.get_offset(get_child_value);
+    if (dense_offset != npos) {
+        SparseState state = sparse_plan.make_state(get_child_value);
+        auto view = input_value.index().create_view(sparse_plan.view_dims);
+        view->lookup(state.lookup_refs);
+        size_t input_subspace;
+        while (view->next_result(state.fetch_addr, input_subspace)) {
+            auto dst = builder->add_subspace(state.output_addr).begin();
+            auto input_offset = input_subspace * dense_plan.in_dense_size;
+            dense_plan.execute(dense_offset,
```

Comment:
you can pass dense_output + input_offset to the plan so you do not have to add it for each cell in the inner loop.